### PR TITLE
RESTWS-900: Get conditions or patient-diagnosis by patient

### PR DESCRIPTION
**Pull request by wikumChamith**
This implements two endpoints that search for the condition of a patient and diagnosis of a patient.

## Description of what I changed

Created search handlers for condition and diagnosis.

## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-900

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by cornel-masson** [comment]
The code looks fine, but I don't see any tests, so I can't verify that it works.

**Comment by wikumChamith** [comment]
> The code looks fine, but I don't see any tests, so I can't verify that it works.

I added a few tests.

**Review by cornel-masson — APPROVED** [review]

**Comment by wikumChamith** [comment]
@dkayiwa can you review this PR?

**Comment by dkayiwa** [comment]
@wikumChamith why is your support starting from platform `2.4` but not before?

**Comment by dkayiwa** [comment]
In which version was that method introduced?

**Comment by wikumChamith** [comment]
> In which version was that method introduced?

Platform `2.3`.

**Comment by dkayiwa** [comment]
How do you tell that? In other words, where do you get that from?

**Comment by wikumChamith** [comment]
> How do you tell that? In other words, where do you get that from?

`ConditionService#getAllConditions` is available from module `omod-2.3`. According to the [documentation](https://docs.openmrs.org/doc/index.html?org/openmrs/api/ConditionService.html), it is available from `2.2.1` .

**Comment by dkayiwa** [comment]
Then why not put it in the `omod-2.2`?

**Comment by wikumChamith** [comment]
> Then why not put it in the `omod-2.2`?

`ConditionService#getAllConditions` is not available for `omod-2.2`. Is there a way I can call that method from `omod-2.2` ?

**Comment by dkayiwa** [comment]
Which exact version is `omod-2.2` referencing?

**Comment by wikumChamith** [comment]
According to `pom.xml`, it is referencing `openmrs.version.2.2.0`.

https://github.com/openmrs/openmrs-module-webservices.rest/blob/9358bda1fa52e36d2949ddae56b820e13ab925bd/omod-2.2/pom.xml#L116-L120

**Comment by dkayiwa** [comment]
What exact value is the `openmrs.version.2.2.0` variable pointing to?

**Comment by wikumChamith** [comment]
> What exact value is the `openmrs.version.2.2.0` variable pointing to?

To 2.2.0-alpha.

https://github.com/openmrs/openmrs-module-webservices.rest/blob/9358bda1fa52e36d2949ddae56b820e13ab925bd/omod-2.2/pom.xml#L13-L15

**Comment by dkayiwa** [comment]
Switch it to `2.2.1`

**Comment by wikumChamith** [comment]
> Switch it to `2.2.1`

I made the necessary changes.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/DiagnosisSearchHandler2_2.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import org.openmrs.module.webservices.rest.web.resource.impl.NeedsPaging;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+import org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_8.PatientResource1_8;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DiagnosisSearchHandler2_2 implements SearchHandler {
+
+	private final SearchConfig searchConfig = new SearchConfig("default", RestConstants.VERSION_1 + "/patientdiagnoses",
+			Collections.singletonList("2.4.* - 9.*"),
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/DiagnosisSearchHandler2_2.java:37]: Is this still starting from 2.4?

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.openmrs.module.webservices.rest.web.resource.impl.NeedsPaging;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+import org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_8.PatientResource1_8;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ConditionSearchHandler2_2 implements SearchHandler {
+
+	private final SearchConfig searchConfig = new SearchConfig("default", RestConstants.VERSION_1 + "/condition",
+			Collections.singletonList("2.2.* - 9.*"),
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:36]: Shouldn't this be from 2.2.1?

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	public PageableResult search(RequestContext context) throws ResponseException {
+		String patientUuid = context.getRequest().getParameter("patientUuid");
+		String includeInactive = context.getRequest().getParameter("includeInactive");
+		ConditionService conditionService = Context.getConditionService();
+		if (patientUuid == null) {
+			return new EmptySearchResult();
+		}
+		Patient patient = ((PatientResource1_8) Context.getService(RestService.class).getResourceBySupportedClass(
+				Patient.class)).getByUniqueId(patientUuid);
+		if (patient == null)
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:59]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/DiagnosisSearchHandler2_2.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	public PageableResult search(RequestContext context) throws ResponseException {
+		String patientUuid = context.getRequest().getParameter("patientUuid");
+		String fromDate = context.getRequest().getParameter("fromDate");
+		if (patientUuid.isEmpty() && fromDate.isEmpty()) {
+			return new EmptySearchResult();
+		}
+		Date dateFrom = (Date) ConversionUtil.convert(fromDate, Date.class);
+		Patient patient = ((PatientResource1_8) Context.getService(RestService.class).getResourceBySupportedClass(
+				Patient.class)).getByUniqueId(patientUuid);
+		if (patient == null)
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/DiagnosisSearchHandler2_2.java:57]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

**Comment by wikumChamith** [comment]
@dkayiwa I fixed the issues mentioned above.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/DiagnosisSearchHandler2_2.java (hunk 1 above)
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/DiagnosisSearchHandler2_2.java]: No, it was a mistake. I changed it to 2.2.1 in the new PR.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java (hunk 2 above)
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java]: yes, I changed it.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import org.openmrs.module.webservices.rest.web.resource.api.SearchParameter;
+import org.openmrs.module.webservices.rest.web.resource.api.SearchQuery;
+import org.openmrs.module.webservices.rest.web.resource.impl.EmptySearchResult;
+import org.openmrs.module.webservices.rest.web.resource.impl.NeedsPaging;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+import org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_8.PatientResource1_8;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ConditionSearchHandler2_2 implements SearchHandler {
```
[the remark this answers is not in the document]
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Did you try overriding the `doSearch()` method?

[the remark this answers is not in the document]
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Yeah, but I couldn't find a way to add extra parameters when using `doSearch()` method.

[the remark this answers is not in the document]
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Is this helpful? https://github.com/openmrs/openmrs-module-webservices.rest/search?q=doSearch

[the remark this answers is not in the document]
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: All these resources only use the default parameter `q` to search or fetch. For `ConditionResource2_2` we need to add one required parameter `patientUuid` and an optional parameter `includeInactive`. So I think we have to implement the SearchHandler interface to create this endpoint.

[the remark this answers is not in the document]
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Did you take a look at this? https://github.com/openmrs/openmrs-module-webservices.rest/blob/2.38.0/omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/ObsResource1_8.java#L474-L506

[the remark this answers is not in the document]
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Obs resource also uses a Search handler. 
https://github.com/openmrs/openmrs-module-webservices.rest/blob/7ec313cb7de970d86c616833af72c772948e8b44/omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ObservationSearchHandler1_8.java#L46-L51

Is it possible to expose new parameters to the Swagger UI by using `doSearch() `method??

[the remark this answers is not in the document]
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Did you try to do as in the link that i gave you?

[the remark this answers is not in the document]
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: yeah. I couldn't find a way to expose parameters to Swagger with that method.

[the remark this answers is not in the document]
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Swagger should not be what determines the implementation.

[the remark this answers is not in the document]
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Should I change this to use `doSearch()` ? What are the benefits of overriding the `doSearch()` over creating a separate search handler class?

[the remark this answers is not in the document]
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java:33]: Simpler and less code to maintain.

[the remark this answers is not in the document]
Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_2/ConditionSearchHandler2_2.java]: I implemented the endpoint by overriding the `doSearch()`.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/DiagnosisResource2_2.java [hunk 6]
```diff
@@ -216,4 +225,20 @@ public Model getUPDATEModel(Representation rep) {
 		        .property("rank", new IntegerProperty())
 		        .property("voided", new BooleanProperty());
 	}
+
+	@Override
+	protected PageableResult doSearch(RequestContext context) {
+		String patientUuid = context.getRequest().getParameter("patientUuid");
+		String fromDate = context.getRequest().getParameter("fromDate");
+		if (patientUuid.isEmpty() && fromDate.isEmpty()) {
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/DiagnosisResource2_2.java:233]: What happens when any of these is null?

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java [hunk 7]
```diff
@@ -224,4 +230,30 @@ public String getDisplayString(Condition condition) {
...
+		ConditionService conditionService = Context.getConditionService();
+		if (patientUuid == null) {
+			return new EmptySearchResult();
+		}
+		Patient patient = ((PatientResource1_8) Context.getService(RestService.class).getResourceBySupportedClass(
+				Patient.class)).getByUniqueId(patientUuid);
+		if (patient == null) {
+			return new EmptySearchResult();
+		}
+		if (includeInactive != null && !includeInactive.isEmpty()) {
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java:247]: StringUtils.isNotBlank()?

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java [hunk 8]
```diff
@@ -224,4 +230,30 @@ public String getDisplayString(Condition condition) {
...
+			boolean isIncludeInactive = Boolean.parseBoolean(includeInactive);
```
Comment (dkayiwa) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java:248]: What will happen to the above code when one supplies `yes` or `no`?

Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java:248]: For all the other inputs other than `true`, `includeInactive` becomes `false`. Should I add a code to handle 'yes/no" like this ?
https://github.com/openmrs/openmrs-module-webservices.rest/blob/4bc9b8a5955d4f373f2784c89d6bccce2db8af27/omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/ObsResource1_8.java#L449-L458

Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java:248]: BooleanUtils.toBoolean?

Comment (wikumChamith) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ConditionResource2_2.java]: I fixed this and a few other issues in the new commit.

**Pull request merged**